pkg/backend/display/richDiff.go

```diff
@@ -0,0 +1,175 @@
+package display
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/engine"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/plugin"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+)
+
+func parseDiffPath(path string) ([]interface{}, error) {
+	// Complete paths obey the following grammar:
+	//
+	// propertyName := [a-zA-Z_$][a-zA-Z0-9_$]*
+	// arrayIndex := '[' [0-9]+ ']'
+	// propertyAccessor := ( '.' propertyName ) | arrayIndex
+	// path := propertyName { propertyAccessor }
+	//
+	// We interpret this a little loosely in order to keep things simple. Specifically, we will accept something close
+	// to the following:
+	// pathElement := { '.' } ( '[' [0-9]+ ']' | [a-zA-Z_$][a-zA-Z0-9_$] )
+	// path := { pathElement }
+
+	var elements []interface{}
+	for len(path) > 0 {
+		switch path[0] {
+		case '.':
+			path = path[1:]
+		case '[':
+			// Look for a closing ']'
+			rbracket := strings.IndexRune(path, ']')
+			if rbracket == -1 {
+				return nil, errors.New("missing ']' in array index")
+			}
+			index, err := strconv.ParseInt(path[1:rbracket], 10, 0)
+			if err != nil {
+				return nil, errors.Wrap(err, "invalid array index")
+			}
+			elements, path = append(elements, int(index)), path[rbracket+1:]
+		default:
+			for i := 0; ; i++ {
+				if i == len(path) || path[i] == '.' || path[i] == '[' {
+					elements, path = append(elements, path[:i]), path[i:]
+					break
+				}
+			}
+		}
+	}
+	return elements, nil
+}
+
+func getProperty(key interface{}, v resource.PropertyValue) resource.PropertyValue {
+	switch {
+	case v.IsArray():
+		index, ok := key.(int)
+		if !ok || index < 0 || index >= len(v.ArrayValue()) {
+			return resource.PropertyValue{}
```

Comment:
should we not error in some way here? (and in the below case as well)?

Reply:
No: the calling code deliberately wants to gracefully handle missing values. The returned `PropertyValue` will be used to determine whether or not a diff is an add or delete if it is `nil` (which it will be in this case).